lib/winston-graylog2.js

```diff
@@ -28,86 +27,62 @@ let getMessageLevel = (function() {
   };
 })();
 
-/**
- * Preparing metadata for graylog
- * If we send a javascript `Error` object
- * to gray log we'll end up with the
- * `[object Object]` string.
- * To have our infos we need to get the stack out of the error.
- * Here we remap metadata to handle this kind of situation
- *
- * @param  {Object} meta
- * @param  {Object} staticMeta
- * @return {Object}
- */
-function prepareMeta(meta, staticMeta) {
-  meta = meta || {};
-
-  if (meta instanceof Error) {
-    meta = {error: meta.stack};
-  } else if (typeof meta === 'object') {
-    meta = _.mapValues(meta, function(value) {
-      if (value instanceof Error) {
-        return value.stack;
-      }
-
-      return value;
+/** Class representing the Graylog2 Winston Transport */
+class Graylog2 extends Transport {
+  /**
+   * Create the transport
+   * @param {Object} options - The options for configuring the transport.
+   */
+  constructor(options) {
+    super(options);
+
+    options = options || {};
+    this.graylog = _.get(options, 'graylog');
+    if (!this.graylog) {
+      this.graylog = {
+        servers: [
+          {
+            host: 'localhost',
+            port: 12201,
+          },
+        ],
+      };
+    }
+
+    this.name = options.name || 'graylog2';
+    this.exceptionsLevel = options.exceptionsLevel || 'not-default';
+
+    this.silent = options.silent || false;
+    this.handleExceptions = options.handleExceptions || false;
+
+    this.staticMeta = options.staticMeta || {};
+
+    this.graylog2 = new Graylog2ClientLibrary(this.graylog);
+
+    this.graylog2.on('error', function(error) {
+      console.error('Error while trying to write to graylog2:', error);
     });
   }
 
-  meta = _.merge(meta, staticMeta);
-
-  return meta;
-}
-
-let Graylog2 = (winston.Graylog2 = function(options) {
-  winston.call(this, options);
-
-  options = options || {};
-  this.graylog = _.get(options, 'graylog');
-  if (!this.graylog) {
-    this.graylog = {
-      servers: [
-        {
-          host: 'localhost',
-          port: 12201,
-        },
-      ],
-    };
+  /**
+   * Log a message to Graylog2.
+   *
+   * @param {Object} info - An object containing the `message` and `info`.
+   * @param {function} callback - Winston's callback to itself.
+   */
+  log(info, callback) {
+    const {message, level} = info;
+    // prettier-ignore
+    this.graylog2[getMessageLevel(level)]( message.substring(0, 100), message, this.staticMeta);
```

Comment:
I'll look more into the implications of both ways and get back to you in the next day or so, but I'll include my reasoning for how I was expecting Error object handling to go under Winston 3 below. I'll also see if I can modify my example project I linked to see how/if it works as I'm anticipating in my below description, since that is something I did not verify.

In the `winston.format` section of the Upgrade Guide I noticed that the handling of Error objects bore a lot of similarity to the old [`rewriters`](https://github.com/winstonjs/winston/blob/master/UPGRADE-3.0.md#rewriters) from older versions of Winston.

So, let's say that you had an error that you wanted to pass into the logger. Your formatter would have to do exactly what `prepareMeta()` used to do: detect that it's been passed an error or object, and construct the output appropriately.

Now, considering that some users of this library may not want to include stack traces in Graylog (they can get big, so lots of them repeatedly could have network and Graylog journaling implications), we should want to give them the option to make a formatter that suits their needs without forcing a bias on them for including a stack trace. **However,** I do think it could be a really good idea to include one or more "suggested formatters" that duplicate what `prepareMeta()` was doing, so that new or refactoring users have the choice of how to deal with such things, as well as having a good example of the kinds of things formatters are capable of doing, and a good base to write their own from. That would also give an upgrade path for refactoring users to get identical expected behavior.

Reply:
Also, regarding your concern about adding `meta` back in, Winston itself doesn't allow that any more. If you want to pass message-specific metadata, then your `meta` needs to be a key on a `message` object, and `message` should be an object instead of a string.

Reply:
Just looked into it a bit more, and [`meta` is now part of the formatters that are available](https://github.com/winstonjs/logform#metadata).

So when constructing your formatter, you would include this one to handle `meta` the way that you're anticipating.

I'll include this handler in the "suggested formatter".

Reply:
I added some more info to the README about how to use the most recent [logform formatters](https://github.com/winstonjs/logform/tree/2.1.0) to get errors and metadata handled correctly.

I also updated the way that log metadata is handled, so that it appears nicely in Graylog when present.

You can see my [updated sample project.](https://github.com/jeremy-j-ackso/wgl-example)